### PR TITLE
feat: ✨track quickfixes [ROAD-688]

### PR DESCRIPTION
**Pull request by bastiandoetsch**

### CODE REVIEW

File: src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt [hunk 1]
```diff
@@ -145,6 +147,22 @@ class SnykAnalyticsService : Disposable {
         }
     }
 
+    fun logQuickFixIsDisplayed(event: QuickFixIsDisplayed) {
+        if (!settings.usageAnalyticsEnabled) return
```
Comment (michelkaporin) [inline src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt:151]: Is `userId.isBlank()` check needed here or userId is always initialised to that point?

File: src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt [hunk 2]
```diff
@@ -7,20 +7,31 @@ import com.intellij.openapi.project.Project
 import com.intellij.openapi.util.TextRange
 import com.intellij.psi.PsiFile
 import com.intellij.util.DocumentUtil
+import io.snyk.plugin.getSnykAnalyticsService
```
Comment (michelkaporin) [inline src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt:10]: Let's fix warnings

**Review by michelkaporin — APPROVED** [review]
Great, left few small comments 👍

File: src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt (hunk 1 above)
Comment (bastiandoetsch) [reply src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt:151]: It is already initialized - you can't get a quickfix without results, and no results without token. And once we get the token, the userID is updated.

File: src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt (hunk 2 above)
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt:10]: I wonder how that snyked in.

**Comment by bastiandoetsch** [comment]
Rebased against master, added changelog, and force-pushed.

File: src/main/kotlin/io/snyk/plugin/Utils.kt [hunk 3]
```diff
@@ -67,6 +68,8 @@ fun getSnykCliDownloaderService(project: Project): SnykCliDownloaderService? = p
 
 fun getSnykProjectSettingsService(project: Project): SnykProjectSettingsStateService? = project.serviceIfNotDisposed()
 
+fun getSnykAnalyticsService(project: Project): SnykAnalyticsService? = project.serviceIfNotDisposed()
```
Comment (ArtsiomCh) [inline src/main/kotlin/io/snyk/plugin/Utils.kt:71]: don't needed
SnykAnalyticsService is not project service

File: src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt [hunk 4]
```diff
@@ -8,19 +8,28 @@ import com.intellij.openapi.util.TextRange
...
+import snyk.analytics.QuickFixIsTriggered
 
 class AlwaysAvailableReplacementIntentionAction(
     val range: TextRange,
     val replacementText: String,
     private val intentionText: String = intentionDefaultTextPrefix,
     private val familyName: String = intentionDefaultFamilyName,
-    val message: String = ""
+    val message: String = "",
+    val analyticsService: SnykAnalyticsService?
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt:22]: could have default value and reassigned just for tests

**Comment by ArtsiomCh** [comment]
Shell we log quickfixes from Container product?

File: src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt (hunk 4 above)
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/common/intentionactions/AlwaysAvailableReplacementIntentionAction.kt:22]: I like that!

**Comment by bastiandoetsch** [comment]
> @bastiandoetsch Please check when it's triggered I can see quickFixIsDisplayed triggered twice even when only 1 popup is shown Also after try to applying it (for package-lock case it doesn't go through), quickFixIsDisplayed triggers multiple times.

Yeah, we cannot control, how often getText() is triggered, or the hover is activated afaik. Do you have any idea?

File: src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt [hunk 5]
```diff
@@ -145,6 +147,22 @@ class SnykAnalyticsService : Disposable {
...
+
+        catchAll(log, "quickFixIsDisplayed") {
+            itly.logQuickFixIsDisplayed(userId, event)
+        }
+    }
+
+    fun logQuickFixIsTriggered(event: QuickFixIsTriggered) {
+        if (!settings.usageAnalyticsEnabled) return
+
+        catchAll(log, "quickFixIsDisplayed") {
```
Comment (ArtsiomCh) [inline src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt:161]: should be `logQuickFixIsTriggered`
PS was confused not to see any logQuickFixIsTriggered in the logs...

**Comment by ArtsiomCh** [comment]
> Yeah, we cannot control, how often getText() is triggered, or the hover is activated afaik. Do you have any idea?
Looks like the only way will be implementing some kind of hacky flag: if it been requested within ~500 ms timeframe then send only one event.
Otherwise we need confirm with Georgy he is ok to have many false events (I don't think so as it ruined the whole idea of such analytic)

File: src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt (hunk 5 above)
Comment (bastiandoetsch) [reply src/main/kotlin/io/snyk/plugin/services/SnykAnalyticsService.kt:161]: Damn copy & paste ;)

**Comment by bastiandoetsch** [comment]
> > Yeah, we cannot control, how often getText() is triggered, or the hover is activated afaik. Do you have any idea?
> 
> Looks like the only way will be implementing some kind of hacky flag: if it been requested within ~500 ms timeframe then send only one event. Otherwise we need confirm with Georgy he is ok to have many false events (I don't think so as it ruined the whole idea of such analytic)

I'd rather not log than using such a hack that can never be precise. @deepcodeg, wdyt?

**Comment by bastiandoetsch** [comment]
Force-push to rebase against master

**Comment by bastiandoetsch** [comment]
If we don't want the `QuickFix is Displayed` event, we can always remove it, according to what Georgi wants. I'll merge this branch now in :).

**Pull request merged**